File: brain/stems.py
"""Vocals-only / instrumental-only stems: classify and pick a two-deck bed.

Canonical scoring lives in `core-rust/clawdj/src/stems.rs` (`clawdj stems pair`).
This module is the in-process planner copy so mix builds do not spawn cargo.
Keep the classify / pair tests in lockstep with the Rust unit tests.
"""
from __future__ import annotations

import json
import re
from typing import Any

from brain.mix_graph import bpm_compatibility, key_compatibility

VOCALS_ONLY = re.compile(
    r"(?:\bacap+ell+a?\b|\ba\s+cappella\b|\bvocals?[- ]only\b)",
    re.IGNORECASE,
)
_INSTRUMENTAL_TAG = re.compile(
    r"(\((?:instrumental|instr\.)\)|\[(?:instrumental|instr\.)\]|\s-\s*instrumental\s*$)",
    re.IGNORECASE,
)
_VERSION_STOP = {
    "instrumental",
    "acapella",
    "acappella",
    "remix",
    "version",
    "dirty",
    "clean",
    "album",
    "feat.",
    "ft.",
    "-",
}


def classify_stem(title: str, path: str = "") -> str:
    hay = f"{title} {path}".lower()
    if VOCALS_ONLY.search(hay):
        return "vocals_only"
    lower_title = title.lower()
    if _INSTRUMENTAL_TAG.search(title) or lower_title.endswith(" instrumental"):
        if "instrumental hip" in lower_title:
            return "full_mix"
        return "instrumental_only"
    if "/instrumental/" in hay:
        return "instrumental_only"
    return "full_mix"


def is_showcase_acapella(notes: str) -> bool:
    return "showcase_acapella" in (notes or "").casefold()


def is_vocal_over_bed(notes: str) -> bool:
    return "vocal_over_bed" in (notes or "").casefold()


def core_title(title: str) -> str:
    stripped = re.sub(r"\([^)]*\)", " ", title or "")
    stripped = re.sub(r"\[[^\]]*\]", " ", stripped)
    return " ".join(
        token
        for token in stripped.lower().split()
        if token and token not in _VERSION_STOP
    )


def pair_vocals(tracks: list[dict[str, Any]]) -> dict[str, Any]:
    used: set[str] = set()
    pairs: list[dict[str, Any]] = []
    claimed: set[str] = set()
    vocals = []
    for vocal_index, vocal in enumerate(tracks):
        kind = classify_stem(vocal.get("title") or "", vocal.get("track_id") or "")
        if kind != "vocals_only":
            continue
        if is_showcase_acapella(vocal.get("dj_notes") or ""):
            continue
        vocals.append((vocal_index, vocal))
    # Neighbor instrumentals first so a later vocal cannot steal this bed.
    leftover = []
    for vocal_index, vocal in vocals:
        neighbor = _neighbor_instrumental(vocal, vocal_index, tracks, used)
        if neighbor is None:
            leftover.append((vocal_index, vocal))
            continue
        used.add(neighbor["bed_id"])
        claimed.add(vocal["track_id"])
        pairs.append(neighbor)
    for vocal_index, vocal in leftover:
        best = _best_bed(vocal, vocal_index, tracks, used)
        if best is None:
            continue
        used.add(best["bed_id"])
        claimed.add(vocal["track_id"])
        pairs.append(best)
    unpaired = [
        vocal["track_id"] for _, vocal in vocals if vocal["track_id"] not in claimed
    ]
    return {"version": 1, "pairs": pairs, "unpaired_vocals": unpaired}


def apply_vocal_layers(tracks: list[dict[str, Any]]) -> tuple[list[dict[str, Any]], list[str]]:
    """Place each dry vocal over a bed: two decks at once, not a transition.

    Mutates a copy of `tracks`. Showcase acapellas are left sequential.
    Vocals that already declare `entry_style=vocal_over_bed` keep their note
    but are still moved so the chosen bed sits immediately before them.
    """
    report = pair_vocals(tracks)
    if report["unpaired_vocals"]:
        by_title = {row["track_id"]: row.get("title") or row["track_id"] for row in tracks}
        names = ", ".join(by_title.get(track_id, track_id) for track_id in report["unpaired_vocals"])
        raise ValueError(
            "vocals-only track(s) have no instrumental bed to layer over: "
            f"{names}. Add an instrumental-only track (or a full mix to loop), "
            "or mark a deliberate showcase_acapella exception"
        )
    notes: list[str] = []
    by_id = {row["track_id"]: dict(row) for row in tracks}
    order = [row["track_id"] for row in tracks]
    for pair in report["pairs"]:
        vocal = by_id[pair["vocal_id"]]
        bed = by_id[pair["bed_id"]]
        vocal_notes = vocal.get("dj_notes") or ""
        if not is_vocal_over_bed(vocal_notes):
            extra = (
                "Two-deck layer, not a mix transition. "
                "entry_style=vocal_over_bed; ride_beats=96; trust_ride_beats; no_flourish"
            )
            if pair["loop_bed"]:
                extra += "; bed_loop_beats=32"
            vocal["dj_notes"] = f"{vocal_notes} {extra}".strip()
        elif pair["loop_bed"] and "bed_loop_beats" not in vocal_notes:
            vocal["dj_notes"] = f"{vocal_notes}; bed_loop_beats=32"
        bed_id = pair["bed_id"]
        vocal_id = pair["vocal_id"]
        order = [item for item in order if item != bed_id]
        vocal_at = order.index(vocal_id)
        order.insert(vocal_at, bed_id)
        notes.append(
            f"layer {vocal.get('title')} over {bed.get('title')} ({pair['reason']})"
        )
    layered = [by_id[track_id] for track_id in order]
    return layered, notes


def assert_vocals_layered(plan: dict[str, Any], notes: dict[str, str] | None = None) -> None:
    """Refuse a built plan that would ride a vocals-only stem by itself.

    Canonical pattern (50centgunitera Get Up over Outta Control Instrumental):
    the instrumental bed stays live, the vocal layers on the other deck,
    `keep_outgoing_live` is set, and there is no solo `play_body` on the
    acapella. A 32-beat dry body is still only vocals.
    """
    notes = notes or {}
    tracks = list(plan.get("tracks") or [])
    by_id = {str(row.get("track_id") or ""): row for row in tracks}
    by_display = {
        f"{row.get('artist', '')} — {row.get('title')}": row for row in tracks
    }
    body_beats = {
        str(event.get("track") or ""): int(event.get("beats") or 0)
        for event in plan.get("events") or []
        if event.get("op") == "play_body"
    }
    layers: dict[str, dict[str, Any]] = {}
    for event in plan.get("events") or []:
        if event.get("technique") != "vocal_over_bed" and "vocal_over_bed" not in (
            event.get("moves") or []
        ):
            continue
        vocal_id = str(event.get("vocal_track_id") or "")
        if not vocal_id:
            incoming = by_display.get(str(event.get("to_track") or ""))
            vocal_id = str((incoming or {}).get("track_id") or "")
        if vocal_id:
            layers[vocal_id] = event
    for track in tracks:
        title = str(track.get("title") or "")
        track_id = str(track.get("track_id") or "")
        if classify_stem(title, track_id) != "vocals_only":
            continue
        note = notes.get(track_id, "") or str(track.get("dj_notes") or "")
        if is_showcase_acapella(note):
            continue
        display = f"{track.get('artist', '')} — {title}"
        solo = body_beats.get(display, 0)
        if solo > 0:
            raise ValueError(
                f"vocals-only track {display} has a solo play_body event "
                f"({solo} beats). Dry vocal is not a mix — pair it with a "
                "compatible instrumental (or a looped instrumental section) "
                "via entry_style=vocal_over_bed, or mark a deliberate "
                "showcase_acapella exception"
            )
        event = layers.get(track_id)
        if event is None:
            raise ValueError(
                f"vocals-only track {display} is not layered over a bed. "
                "Use entry_style=vocal_over_bed on a two-deck pair, or mark "
                "showcase_acapella"
            )
        if not event.get("keep_outgoing_live"):
            raise ValueError(
                f"vocals-only track {display} must keep the instrumental bed "
                "live (keep_outgoing_live). Fade the vocal out; do not stop "
                "the bed — Get Up (Acapella) over Outta Control Instrumental "
                "is the pattern"
            )
        bed_id = str(event.get("bed_track_id") or "")
        bed = by_id.get(bed_id)
        if bed is None:
            bed = by_display.get(str(event.get("from_track") or ""))
        bed_kind = classify_stem(
            str((bed or {}).get("title") or ""),
            str((bed or {}).get("track_id") or bed_id),
        )
        moves = event.get("moves") or []
        if bed_kind == "instrumental_only" or "bed_loop" in moves:
            continue
        raise ValueError(
            f"vocals-only track {display} is layered over a full mix with "
            "vocals. Use an instrumental-only bed, or loop a cued "
            "instrumental section (bed_loop_beats / bed_loop)"
        )


def _neighbor_instrumental(
    vocal: dict[str, Any],
    vocal_index: int,
    tracks: list[dict[str, Any]],
    used: set[str],
) -> dict[str, Any] | None:
    """An instrumental sitting next to the acapella is the intended bed."""
    for delta in (-1, 1):
        index = vocal_index + delta
        if index < 0 or index >= len(tracks):
            continue
        bed = tracks[index]
        if bed.get("track_id") in used:
            continue
        kind = classify_stem(bed.get("title") or "", bed.get("track_id") or "")
        if kind == "instrumental_only":
            return _score_bed(vocal, vocal_index, bed, index, kind)
    return None


def _best_bed(
    vocal: dict[str, Any],
    vocal_index: int,
    tracks: list[dict[str, Any]],
    used: set[str],
) -> dict[str, Any] | None:
    best: dict[str, Any] | None = None
    for bed_index, bed in enumerate(tracks):
        if bed.get("track_id") == vocal.get("track_id"):
            continue
        if bed.get("track_id") in used:
            continue
        kind = classify_stem(bed.get("title") or "", bed.get("track_id") or "")
        if kind == "vocals_only":
            continue
        pair = _score_bed(vocal, vocal_index, bed, bed_index, kind)
        if best is None or pair["score"] > best["score"] or (
            pair["score"] == best["score"] and pair["bed_id"] < best["bed_id"]
        ):
            best = pair
    return best


def _score_bed(
    vocal: dict[str, Any],
    vocal_index: int,
    bed: dict[str, Any],
    bed_index: int,
    kind: str,
) -> dict[str, Any]:
    same_song = bool(core_title(vocal.get("title") or "")) and (
        core_title(vocal.get("title") or "") == core_title(bed.get("title") or "")
    )
    adjacent = abs(vocal_index - bed_index) == 1
    score = 0.0
    reasons: list[str] = []
    if kind == "instrumental_only":
        score += 3.0
        reasons.append("instrumental-only bed")
    else:
        score += 0.5
        reasons.append("full-mix looped section bed")
    bpm_score, bpm_reason = bpm_compatibility(vocal.get("bpm"), bed.get("bpm"))
    score += float(bpm_score) * 2.0
    if bpm_reason:
        reasons.append(bpm_reason)
    key_score, key_reason = key_compatibility(vocal.get("key"), bed.get("key"))
    score += float(key_score) * 2.0
    if key_reason:
        reasons.append(key_reason)
    if same_song:
        reasons.append("same-song stack (legal, not preferred)")
    elif key_score >= 0.55:
        score += 2.0
        reasons.append("different song")
    if adjacent:
        score += 1.0
        reasons.append("already adjacent")
    return {
        "vocal_id": vocal["track_id"],
        "bed_id": bed["track_id"],
        "score": score,
        "reason": "; ".join(reasons),
        "same_song": same_song,
        "bed_kind": kind,
        "loop_bed": kind == "full_mix",
    }


def dumps_pair_payload(tracks: list[dict[str, Any]]) -> str:
    """JSON array `clawdj stems pair` accepts on stdin."""
    payload = []
    for row in tracks:
        payload.append(
            {
                "track_id": row["track_id"],
                "title": row.get("title") or "",
                "artist": row.get("artist") or "",
                "bpm": row.get("bpm"),
                "key": row.get("key"),
                "path": row.get("track_id"),
                "dj_notes": row.get("dj_notes") or "",
            }
        )
    return json.dumps(payload)

File: core-rust/clawdj/src/stems.rs
//! Classify vocals-only / instrumental-only stems and pick a bed for a dry vocal.
//!
//! Acapellas are not songs. They play at the same time as a beat-matched
//! instrumental (or a looped instrumental section of a full mix). Same-song
//! instrumental stacks are legal but not preferred — that stack is the original.

use serde::{Deserialize, Serialize};

/// How a track should be treated when planning a mix.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum StemKind {
    VocalsOnly,
    InstrumentalOnly,
    FullMix,
}

/// One playlist row the pairer can see.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct StemTrack {
    pub track_id: String,
    pub title: String,
    #[serde(default)]
    pub artist: String,
    #[serde(default)]
    pub bpm: Option<f64>,
    #[serde(default)]
    pub key: Option<String>,
    #[serde(default)]
    pub path: Option<String>,
    /// Effective dj_notes. `showcase_acapella` opts out of forced layering.
    #[serde(default)]
    pub dj_notes: String,
}

/// One vocals-only track assigned to a bed.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct StemPair {
    pub vocal_id: String,
    pub bed_id: String,
    pub score: f64,
    pub reason: String,
    pub same_song: bool,
    pub bed_kind: StemKind,
    pub loop_bed: bool,
}

/// JSON report for `clawdj stems pair`.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct StemPairReport {
    pub version: u32,
    pub pairs: Vec<StemPair>,
    pub unpaired_vocals: Vec<String>,
}

/// Classify a title/path. Version tags win; "Vocal Remix" of a produced
/// record is a full mix unless it is also tagged acapella.
#[must_use]
pub fn classify_stem(title: &str, path: &str) -> StemKind {
    let hay = format!("{title} {path}").to_ascii_lowercase();
    if is_vocals_only(&hay) {
        StemKind::VocalsOnly
    } else if is_instrumental_only(title, &hay) {
        StemKind::InstrumentalOnly
    } else {
        StemKind::FullMix
    }
}

/// True when a human note opts this vocal out of forced layering.
#[must_use]
pub fn is_showcase_acapella(notes: &str) -> bool {
    notes.to_ascii_lowercase().contains("showcase_acapella")
}

/// True when the vocal is already planned as a two-deck layer.
#[must_use]
pub fn is_vocal_over_bed(notes: &str) -> bool {
    notes.to_ascii_lowercase().contains("vocal_over_bed")
}

/// Pair each non-showcase vocals-only track with the best unused bed.
/// Prefers an instrumental-only track already in the set; a full mix is a
/// last-resort looped-section bed. Same-song instrumentals score lower than
/// a different interesting instrumental.
#[must_use]
pub fn pair_vocals(tracks: &[StemTrack]) -> StemPairReport {
    let mut used_beds = Vec::new();
    let mut pairs = Vec::new();
    let mut vocals = Vec::new();
    for (vocal_index, vocal) in tracks.iter().enumerate() {
        let path = vocal.path.as_deref().unwrap_or(vocal.track_id.as_str());
        if classify_stem(&vocal.title, path) != StemKind::VocalsOnly {
            continue;
        }
        if is_showcase_acapella(&vocal.dj_notes) {
            continue;
        }
        vocals.push((vocal_index, vocal));
    }
    let mut leftover = Vec::new();
    for &(vocal_index, vocal) in &vocals {
        if let Some(pair) = neighbor_instrumental(vocal, vocal_index, tracks, &used_beds) {
            used_beds.push(pair.bed_id.clone());
            pairs.push(pair);
        } else {
            leftover.push((vocal_index, vocal));
        }
    }
    for (vocal_index, vocal) in leftover {
        if let Some(pair) = best_bed(vocal, vocal_index, tracks, &used_beds) {
            used_beds.push(pair.bed_id.clone());
            pairs.push(pair);
        }
    }
    let claimed: std::collections::HashSet<&str> =
        pairs.iter().map(|pair| pair.vocal_id.as_str()).collect();
    let unpaired = vocals
        .iter()
        .filter(|(_, vocal)| !claimed.contains(vocal.track_id.as_str()))
        .map(|(_, vocal)| vocal.track_id.clone())
        .collect();
    StemPairReport {
        version: 1,
        pairs,
        unpaired_vocals: unpaired,
    }
}

fn is_vocals_only(hay: &str) -> bool {
    hay.contains("acappella")
        || hay.contains("acapella")
        || hay.contains("a cappella")
        || hay.contains("a capella")
        || hay.contains("vocals only")
        || hay.contains("vocal only")
}

fn is_instrumental_only(title: &str, hay: &str) -> bool {
    let lower_title = title.to_ascii_lowercase();
    lower_title.contains("(instrumental)")
        || lower_title.contains("[instrumental]")
        || lower_title.contains("(instr.)")
        || lower_title.ends_with(" - instrumental")
        || hay.contains("/instrumental/")
        || version_tag_instrumental(&lower_title)
}

fn version_tag_instrumental(title: &str) -> bool {
    title.ends_with(" instrumental") && !title.contains("instrumental hip")
}

fn neighbor_instrumental(
    vocal: &StemTrack,
    vocal_index: usize,
    tracks: &[StemTrack],
    used_beds: &[String],
) -> Option<StemPair> {
    for delta in [-1_isize, 1] {
        let index = isize::try_from(vocal_index).ok()?.checked_add(delta)?;
        if index < 0 {
            continue;
        }
        let bed_index = usize::try_from(index).ok()?;
        let Some(bed) = tracks.get(bed_index) else {
            continue;
        };
        if used_beds.iter().any(|id| id == &bed.track_id) {
            continue;
        }
        let path = bed.path.as_deref().unwrap_or(bed.track_id.as_str());
        if classify_stem(&bed.title, path) == StemKind::InstrumentalOnly {
            return Some(score_bed(
                vocal,
                vocal_index,
                bed,
                bed_index,
                StemKind::InstrumentalOnly,
            ));
        }
    }
    None
}

fn best_bed(
    vocal: &StemTrack,
    vocal_index: usize,
    tracks: &[StemTrack],
    used_beds: &[String],
) -> Option<StemPair> {
    let mut best: Option<StemPair> = None;
    for (bed_index, bed) in tracks.iter().enumerate() {
        if bed.track_id == vocal.track_id {
            continue;
        }
        if used_beds.iter().any(|id| id == &bed.track_id) {
            continue;
        }
        let path = bed.path.as_deref().unwrap_or(bed.track_id.as_str());
        let kind = classify_stem(&bed.title, path);
        if kind == StemKind::VocalsOnly {
            continue;
        }
        let pair = score_bed(vocal, vocal_index, bed, bed_index, kind);
        let take = match &best {
            None => true,
            Some(current) => {
                pair.score > current.score
                    || (pair.score == current.score && pair.bed_id < current.bed_id)
            }
        };
        if take {
            best = Some(pair);
        }
    }
    best
}

fn score_bed(
    vocal: &StemTrack,
    vocal_index: usize,
    bed: &StemTrack,
    bed_index: usize,
    kind: StemKind,
) -> StemPair {
    let same_song =
        core_title(&vocal.title) == core_title(&bed.title) && !core_title(&vocal.title).is_empty();
    let adjacent = vocal_index.abs_diff(bed_index) == 1;
    let mut score = 0.0;
    let mut reasons = Vec::new();
    if kind == StemKind::InstrumentalOnly {
        score += 3.0;
        reasons.push("instrumental-only bed".to_string());
    } else {
        score += 0.5;
        reasons.push("full-mix looped section bed".to_string());
    }
    let (bpm_score, bpm_reason) = bpm_compatibility(vocal.bpm, bed.bpm);
    score += bpm_score * 2.0;
    reasons.push(bpm_reason);
    let (key_score, key_reason) = key_compatibility(vocal.key.as_deref(), bed.key.as_deref());
    score += key_score * 2.0;
    reasons.push(key_reason);
    if same_song {
        reasons.push("same-song stack (legal, not preferred)".to_string());
    } else if key_score >= 0.55 {
        score += 2.0;
        reasons.push("different song".to_string());
    }
    if adjacent {
        score += 1.0;
        reasons.push("already adjacent".to_string());
    }
    StemPair {
        vocal_id: vocal.track_id.clone(),
        bed_id: bed.track_id.clone(),
        score,
        reason: reasons.join("; "),
        same_song,
        bed_kind: kind,
        loop_bed: kind == StemKind::FullMix,
    }
}

fn core_title(title: &str) -> String {
    let mut stripped = title.to_string();
    while let Some(start) = stripped.find('(') {
        if let Some(end) = stripped[start..].find(')') {
            stripped.replace_range(start..start + end + 1, " ");
        } else {
            break;
        }
    }
    while let Some(start) = stripped.find('[') {
        if let Some(end) = stripped[start..].find(']') {
            stripped.replace_range(start..start + end + 1, " ");
        } else {
            break;
        }
    }
    stripped
        .split_whitespace()
        .filter(|word| {
            let lower = word.to_ascii_lowercase();
            !matches!(
                lower.as_str(),
                "instrumental"
                    | "acapella"
                    | "acappella"
                    | "remix"
                    | "version"
                    | "dirty"
                    | "clean"
                    | "album"
                    | "feat."
                    | "ft."
                    | "-"
            )
        })
        .map(str::to_ascii_lowercase)
        .collect::<Vec<_>>()
        .join(" ")
}

fn bpm_compatibility(a: Option<f64>, b: Option<f64>) -> (f64, String) {
    match (a, b) {
        (Some(a), Some(b)) if a > 0.0 && b > 0.0 => {
            let lo = a.min(b);
            let hi = a.max(b);
            for factor in [1.0_f64, 2.0, 0.5] {
                let mut ratio = hi / (lo * factor);
                if ratio < 1.0 {
                    ratio = 1.0 / ratio;
                }
                if ratio <= 1.02 {
                    return (1.0, format!("bpm nearly identical ({a:.1}↔{b:.1})"));
                }
                if ratio <= 1.08 {
                    return (0.9, format!("bpm within ~8% ({a:.1}↔{b:.1})"));
                }
            }
            (0.15, format!("bpm far ({a:.1}↔{b:.1})"))
        }
        _ => (0.35, "bpm unknown".to_string()),
    }
}

fn key_compatibility(a: Option<&str>, b: Option<&str>) -> (f64, String) {
    let (Some(ka), Some(kb)) = (
        parse_camelot(a.unwrap_or("")),
        parse_camelot(b.unwrap_or("")),
    ) else {
        return (0.4, "key unknown".to_string());
    };
    if ka == kb {
        return (1.0, format!("same key ({})", a.unwrap_or("")));
    }
    if ka.0 == kb.0 {
        return (
            0.95,
            format!(
                "relative major/minor ({}↔{})",
                a.unwrap_or(""),
                b.unwrap_or("")
            ),
        );
    }
    let diff = (i32::from(ka.0) - i32::from(kb.0)).unsigned_abs() % 12;
    let wrapped = diff.min(12 - diff);
    if wrapped == 1 && ka.1 == kb.1 {
        return (
            0.85,
            format!("camelot neighbor ({}↔{})", a.unwrap_or(""), b.unwrap_or("")),
        );
    }
    (
        0.15,
        format!("key clash ({}↔{})", a.unwrap_or(""), b.unwrap_or("")),
    )
}

fn parse_camelot(key: &str) -> Option<(u8, char)> {
    let raw = key.trim().replace(' ', "");
    if raw.is_empty() {
        return None;
    }
    if let Some(num) = raw
        .strip_suffix(['A', 'a', 'B', 'b'])
        .and_then(|digits| digits.parse::<u8>().ok())
    {
        if (1..=12).contains(&num) {
            let mode = raw.chars().last()?.to_ascii_uppercase();
            return Some((num, mode));
        }
    }
    let lowered = raw.to_ascii_lowercase();
    let minor = lowered.ends_with('m') || lowered.ends_with("min");
    let note = raw
        .trim_end_matches(|c: char| {
            c.eq_ignore_ascii_case(&'m')
                || c.eq_ignore_ascii_case(&'i')
                || c.eq_ignore_ascii_case(&'n')
        })
        .to_ascii_uppercase();
    let num = if minor {
        camelot_minor(&note)
    } else {
        camelot_major(&note)
    }?;
    Some((num, if minor { 'A' } else { 'B' }))
}

fn camelot_major(note: &str) -> Option<u8> {
    Some(match note {
        "C" => 8,
        "G" => 9,
        "D" => 10,
        "A" => 11,
        "E" => 12,
        "B" => 1,
        "F#" | "GB" => 2,
        "DB" | "C#" => 3,
        "AB" | "G#" => 4,
        "EB" | "D#" => 5,
        "BB" | "A#" => 6,
        "F" => 7,
        _ => return None,
    })
}

fn camelot_minor(note: &str) -> Option<u8> {
    Some(match note {
        "A" => 8,
        "E" => 9,
        "B" => 10,
        "F#" | "GB" => 11,
        "C#" | "DB" => 12,
        "G#" | "AB" => 1,
        "D#" | "EB" => 2,
        "A#" | "BB" => 3,
        "F" => 4,
        "C" => 5,
        "G" => 6,
        "D" => 7,
        _ => return None,
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    fn track(id: &str, title: &str, bpm: f64, key: &str) -> StemTrack {
        StemTrack {
            track_id: id.to_string(),
            title: title.to_string(),
            artist: "50 Cent".to_string(),
            bpm: Some(bpm),
            key: Some(key.to_string()),
            path: Some(id.to_string()),
            dj_notes: String::new(),
        }
    }

    #[test]
    fn acapella_is_vocals_only_vocal_remix_is_not() {
        assert_eq!(
            classify_stem(
                "Get Up (Acapella)",
                "/x/04 - 50 Cent - Get Up (Acapella).mp3"
            ),
            StemKind::VocalsOnly
        );
        assert_eq!(
            classify_stem("Jimmy Crack Corn (Vocal Remix)", "/x/vocal-remix.mp3"),
            StemKind::FullMix
        );
        assert_eq!(
            classify_stem(
                "Jimmy Crack Corn (Vocal Remix) (Feat. Cashis) (Acapella)",
                "/x/a.mp3"
            ),
            StemKind::VocalsOnly
        );
    }

    #[test]
    fn dashed_instrumental_title_is_a_stem() {
        assert_eq!(
            classify_stem("Outta Control - Instrumental", "/x/outta.mp3"),
            StemKind::InstrumentalOnly
        );
        assert_eq!(
            classify_stem("Get Up (Instrumental)", "/x/get-up-inst.mp3"),
            StemKind::InstrumentalOnly
        );
    }

    #[test]
    fn neighbor_instrumental_wins_over_a_distant_mash() {
        let tracks = vec![
            track(
                "/jcc-inst",
                "Jimmy Crack Corn (Vocal Remix) (Instrumental)",
                95.86,
                "Fm",
            ),
            track(
                "/jcc-acapella",
                "Jimmy Crack Corn (Vocal Remix) (Feat. Cashis) (Acapella)",
                97.0,
                "Fm",
            ),
            track(
                "/jcc-full",
                "Eminem & 50 Cent - Jimmy Crack Corn",
                95.86,
                "Fm",
            ),
            track("/hands-up-inst", "Hands Up (Instrumental)", 95.0, "Fm"),
        ];
        let report = pair_vocals(&tracks);
        assert_eq!(report.pairs[0].bed_id, "/jcc-inst");
    }

    #[test]
    fn get_up_prefers_adjacent_different_song_instrumental() {
        let tracks = vec![
            track("/best-friend", "Best Friend", 90.8, "Db"),
            track("/get-up-acapella", "Get Up (Acapella)", 186.0, "F#"),
            track("/outta-inst", "Outta Control - Instrumental", 92.0, "Ebm"),
            track("/get-up-inst", "Get Up (Instrumental)", 93.0, "F#"),
        ];
        let report = pair_vocals(&tracks);
        assert_eq!(report.pairs.len(), 1);
        assert_eq!(report.pairs[0].vocal_id, "/get-up-acapella");
        assert_eq!(report.pairs[0].bed_id, "/outta-inst");
        assert!(!report.pairs[0].same_song);
        assert!(!report.pairs[0].loop_bed);
    }

    #[test]
    fn vocal_over_bed_note_is_detected() {
        assert!(is_vocal_over_bed(
            "entry_style=vocal_over_bed; ride_beats=96"
        ));
        assert!(!is_vocal_over_bed("showcase_acapella"));
    }

    #[test]
    fn showcase_acapella_is_not_paired() {
        let mut vocal = track("/banks", "On Fire (Acapella)", 94.0, "F#m");
        vocal.dj_notes = "showcase_acapella".to_string();
        let tracks = vec![vocal, track("/inst", "On Fire (Instrumental)", 94.0, "F#m")];
        let report = pair_vocals(&tracks);
        assert!(report.pairs.is_empty());
    }

    #[test]
    fn key_clash_does_not_beat_same_song_instrumental() {
        let tracks = vec![
            track("/cream-acapella", "C.R.E.A.M. (A Cappella)", 141.0, "Ab"),
            track("/cream-inst", "C.R.E.A.M. (Instrumental)", 93.0, "Ab"),
            track(
                "/patiently-inst",
                "Patiently Waiting (Instrumental)",
                79.0,
                "Bm",
            ),
        ];
        let report = pair_vocals(&tracks);
        assert_eq!(report.pairs[0].bed_id, "/cream-inst");
    }

    #[test]
    fn already_noted_vocal_still_pairs_onto_its_bed() {
        let mut vocal = track(
            "/still-kill-acapella",
            "I'll Still Kill (Acappella)",
            110.0,
            "F",
        );
        vocal.dj_notes = "entry_style=vocal_over_bed; ride_beats=192".to_string();
        let tracks = vec![
            track(
                "/still-kill-inst",
                "I'll Still Kill (Instrumental)",
                88.0,
                "Dm",
            ),
            vocal,
        ];
        let report = pair_vocals(&tracks);
        assert_eq!(report.pairs.len(), 1);
        assert_eq!(report.pairs[0].bed_id, "/still-kill-inst");
    }

    #[test]
    fn already_layered_neighbor_is_not_stolen() {
        let mut jcc = track(
            "/jcc-acapella",
            "Jimmy Crack Corn (Vocal Remix) (Feat. Cashis) (Acapella)",
            97.0,
            "Fm",
        );
        jcc.dj_notes = "entry_style=vocal_over_bed; ride_beats=96".to_string();
        let tracks = vec![
            track(
                "/jcc-inst",
                "Jimmy Crack Corn (Vocal Remix) (Instrumental)",
                95.86,
                "Fm",
            ),
            jcc,
            track("/ydk-acapella", "You Don't Know [Acapella]", 85.7, "Ab"),
            track("/ydk-inst", "You Don't Know [Instrumental]", 85.7, "Ab"),
        ];
        let report = pair_vocals(&tracks);
        let jcc_bed = report
            .pairs
            .iter()
            .find(|pair| pair.vocal_id == "/jcc-acapella")
            .unwrap();
        let ydk_bed = report
            .pairs
            .iter()
            .find(|pair| pair.vocal_id == "/ydk-acapella")
            .unwrap();
        assert_eq!(jcc_bed.bed_id, "/jcc-inst");
        assert_eq!(ydk_bed.bed_id, "/ydk-inst");
    }
}
